x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java

```diff
@@ -110,13 +122,53 @@ public void registerGeoShapeCentroidAggregator(ValuesSourceRegistry.Builder buil
             });
     }
 
-    public static void registerValueCountAggregator(ValuesSourceRegistry.Builder builder) {
+    private void registerGeoShapeGridAggregators(ValuesSourceRegistry.Builder builder) {
+        builder.register(GeoHashGridAggregationBuilder.NAME, GeoShapeValuesSourceType.instance(),
+            (GeoGridAggregatorSupplier) (name, factories, valuesSource, precision, geoBoundingBox, requiredSize, shardSize,
+                                         aggregationContext, parent, metadata) -> {
+                if (getLicenseState().isAllowed(XPackLicenseState.Feature.SPATIAL_GEO_GRID)) {
+                    final GeoGridTiler tiler;
+                    if (geoBoundingBox.isUnbounded()) {
+                        tiler = new GeoHashGridTiler();
+                    } else {
+                        tiler = new BoundedGeoHashGridTiler(geoBoundingBox);
+                    }
+                    GeoShapeCellIdSource cellIdSource = new GeoShapeCellIdSource((GeoShapeValuesSource) valuesSource, precision, tiler);
+                    GeoShapeHashGridAggregator agg = new GeoShapeHashGridAggregator(name, factories, cellIdSource, requiredSize, shardSize,
+                        aggregationContext, parent, metadata);
+                    cellIdSource.setCircuitBreakerConsumer(agg::addRequestBytes);
```

Comment:
I think I figured out the circuit-breaking logic!

this is my least favorite part. a hack to give GeoShapeCellIdSource access to AggregatorBase's circuit-breaker logic, which must be created before the aggregator is created.

Reply:
I cannot think in another way to achieve this at the moment. Do you see any issue down the line with this approach. Maybe add a TODO so we do not forget we should find a better solution?

I wonder if in the long term we need a family of value source that knows about the circuit breaker.

Reply:
> Do you see any issue down the line with this approach

I do not see a way around this in the short-term. I think it would require more changes to more inherited methods to properly pass along this information from the aggregator to the circuit-breaker via the ValuesSource.

> I wonder if in the long term we need a family of value source that knows about the circuit breaker.

this was the approach taken in #50720. I realize the problem there is that the Aggregator has its own utility method for keeping track of bytes and then subtracting from the circuit-breaker upon closing. I had some bugs with separating the concerns and so I updated it to leverage the Aggregator's wrapper around circuit-breaking logic.

I will add a comment in the code for sure